pkg/render/applicationlayer.go

```diff
@@ -0,0 +1,308 @@
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/secret"
+	"github.com/tigera/operator/pkg/render/common/statik"
+)
+
+const (
+	CalicoSystemNamespace       = "calico-system"
+	EnvoyConfigConfigMapName    = "envoy-config"
+	L7LogCollectorDeamonsetName = "l7-log-collector"
+	L7CollectorContainerName    = "l7-collector"
+	ProxyContainerName          = "envoy-proxy"
+	EnvoyLogsKey                = "envoy-logs"
+	EnvoyConfigKey              = "envoy-config"
+	EnvoyConfigMapKey           = "envoy-config.yaml"
+)
+
+func ApplicationLayer(pullSecrets []*corev1.Secret, installation *operatorv1.InstallationSpec, osType rmeta.OSType,
+	applicationLayer *operatorv1.ApplicationLayer, l7Spec *L7LogCollectionSpec) Component {
+
+	return &ApplicationLayerComponent{
+		pullSecrets:      pullSecrets,
+		installation:     installation,
+		osType:           osType,
+		applicationLayer: applicationLayer,
+		l7Spec:           l7Spec,
+	}
+}
+
+type L7LogCollectionSpec struct {
+	Enabled                bool
+	LogIntervalSeconds     *int64
+	LogRequestsPerInterval *int64
+}
+
+type ApplicationLayerComponent struct {
+	pullSecrets      []*corev1.Secret
+	installation     *operatorv1.InstallationSpec
+	osType           rmeta.OSType
+	applicationLayer *operatorv1.ApplicationLayer
+	l7Spec           *L7LogCollectionSpec
+	proxyImage       string
+	collectorImage   string
+}
+
+func (c *ApplicationLayerComponent) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := c.installation.Registry
+	path := c.installation.ImagePath
+	prefix := c.installation.ImagePrefix
+
+	if c.osType != c.SupportedOSType() {
+		return fmt.Errorf("l7 log collection is supported only on %s", c.SupportedOSType())
+	}
+
+	var err error
+	var errMsgs []string
+
+	c.proxyImage, err = components.GetReference(components.ComponentEnvoyProxy, reg, path, prefix, is)
+
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+	c.collectorImage, err = components.GetReference(components.ComponentL7Collector, reg, path, prefix, is)
+
+	if err != nil {
+		errMsgs = append(errMsgs, err.Error())
+	}
+
+	if len(errMsgs) != 0 {
+		return fmt.Errorf(strings.Join(errMsgs, ","))
+	}
+
+	return nil
+}
+
+func (c *ApplicationLayerComponent) SupportedOSType() rmeta.OSType {
+	return rmeta.OSTypeLinux
+}
+
+func (c *ApplicationLayerComponent) Objects() ([]client.Object, []client.Object) {
+	var objs []client.Object
+	// TODO: ensure that namespace exists before proceeding
+	// if l7spec is provided render the required objects
+	if c.l7Spec != nil && c.l7Spec.Enabled {
+		objs = append(objs, c.envoyL7ConfigMap())
+		objs = append(objs, c.daemonset())
+	}
+
+	return objs, nil
+}
+
+func (c *ApplicationLayerComponent) Ready() bool {
+	return true
+}
+
+// daemonset creates a daemonset for the L7 log collector component.
+func (c *ApplicationLayerComponent) daemonset() *appsv1.DaemonSet {
+	maxUnavailable := intstr.FromInt(1)
+
+	annots := map[string]string{}
```

Comment:
We should have an annotation for the configmap so if it changes in the future the daemonset will be updated.